### ISSUE TITLE
Controller.saveViewState causes problems with nested Controllers and common view IDs

### ISSUE THREAD

**Issue by jhorvat** [bluelinelabs/Conductor#346]
I'm running into an issue where my app crashes on rotation with:

```
java.lang.IllegalArgumentException: Wrong state class, expecting View State but received class android.support.v7.widget.RecyclerView$SavedState instead. This usually happens when two views of different type have the same id in the same hierarchy.
```

The view hierarchy has a root controller with a viewpager that is hosting controllers of its own. The first child controller hosts a custom ViewGroup that wraps a RecyclerView but is treated as if it is a RecyclerView by callers. The second controller hosts a plain old RecyclerView. Both of these views have an id of ```recycler_view```. Obviously renaming the ID of one of these will solve the bug but I think that there's still a problem with how nested Controller's view state is saved that is causing this issue.

```
--------------------------------------------
| Parent Controller                        |
|  --------------------------------------  |
|  |    Child    |    Child             |  |
|  |  Controller |  Controller          |  |
|  |      1      |      2               |  |
|  --------------------------------------  |
--------------------------------------------
```

Currently when controllers are saved, ```saveHierarchyState``` is called on the controller's root view which recurses the view tree under said controller.  This works great for Child Controllers 1 and 2 in the diagram but when Parent Controller saves its hierarchy this includes the views under its children's control. If those children have conflicting view IDs then the saved state of 2 will overwrite 1 producing the error I experienced above. It also means that save/restore work is duplicated since the Child and Parent Controllers call the same save/restore methods on the views belonging to the children. 

I've created a repro of the problem [here](https://github.com/jhorvat/ViewInstanceStateRepro). The crash in the repro can be avoided by either changing the ID of one of the views with ```recycler_view``` as their ID or swapping the order of ```frame_container``` and ```recycler_container``` in the ```LinearLayout```.

**Comment by EricKuck** [comment]
Is there any proposed solution to this issue? Outside of `saveHierarchyState`, I'm not aware of methods that will properly save a controller's state. The only thing I can think of would be to post-process the result of `saveHierarchyState` in order to remove entries that are from child controllers. That would be very tricky if we have deeply nested controllers though.

**Comment by jhorvat** [comment]
@EricKuck yeah I'm not sure post-processing would even work in this situation since the root cause is data being overwritten during the save process. I'm not super familiar with the deep inner workings of Conductor, but I guess that maybe reimplementing the ```saveHierarchyState``` implementation to respect ```ViewGroup```s being managed by a child ```Controller``` i.e. not traversing those groups. 

I wonder how AOSP handles this with respect to Activities and nested Fragments because I don't believe this issue exists there.

**Comment by PaulWoitaschek** [comment]
Iterating through the views and using bundles for each controller hosted view could work. 

I faced that issue too some time ago and it actually took some time to find all the edge cases where this was happening in my code. 
My workaround is to give each child really unique IDs. The thing here is that its not so important with these crashes as you actually notice the failure. But if it's the same view type, you end up with wrong view States but you never notice it.

**Comment by jhorvat** [comment]
Yeah the incorrect view states is also an unfortunate bug. Unique IDs aren't always feasible unfortunately, for instance the app I am using Conductor on has ```ViewPager```s with the same Controllers but different data all over the place, generating unique IDs on the fly for those pages would be really inconvenient/messy. For now I've set my custom ```ViewGroup``` to avoid the crash by overriding ```onRestoreInstanceState```

```java
   @Override
   protected void onRestoreInstanceState(Parcelable state) {
        if (state instanceof android.support.v4.view.AbsSavedState) {
            super.onRestoreInstanceState(((AbsSavedState) state).getSuperState());
        } else {
            super.onRestoreInstanceState(state);
        }
    }
```

**Comment by lykhonis** [comment]
A solution is to call [View.setSaveFromParentEnabled(false)](https://developer.android.com/reference/android/view/View.html#setSaveFromParentEnabled(boolean)) on a View from a Controller once one is created.

**Comment by jhorvat** [comment]
@VolodymyrLykhonis so essentially any child controller should call ```setSaveFromParentEnabled(false)``` on its root view?

**Comment by lykhonis** [comment]
@jhorvat yes, basically once a controller gets its view, it should call this method. This is a one liner fix in Conductor.
